src/gConsent/query/query.ts

```diff
@@ -0,0 +1,262 @@
+   */
+  prev?: CredentialFilter;
+  /**
+   * Next page of query results.
+   */
+  next?: CredentialFilter;
+  /**
+   * Last page of query results.
+   */
+  last?: CredentialFilter;
+};
+
+function toCredentialFilter(url: URL): CredentialFilter {
+  const filter: CredentialFilter = {};
+  url.searchParams.forEach((value, key) => {
+    if (isSupportedFilterElement(key)) {
+      Object.assign(filter, { [`${key}`]: value });
+    }
+  });
+  return filter;
+}
+
+function hasItems(x: unknown): x is { items: unknown[] } {
+  const candidate = x as { items: unknown };
+  return candidate.items !== undefined && Array.isArray(candidate.items);
+}
+
+function isObjectWithId(x: unknown): x is { id: string } {
+  const candidate = x as { id: string };
+  return typeof candidate.id === "string";
+}
+
+async function toCredentialResult(
+  response: Response,
+): Promise<CredentialResult> {
+  const result: CredentialResult = { items: [] };
+  const linkHeader = response.headers.get("Link");
+  if (linkHeader !== null) {
+    const parsedLinks = LinkHeader.parse(linkHeader);
+    PAGING_RELS.forEach((rel) => {
+      const link = parsedLinks.get("rel", rel);
+      if (link.length > 1) {
+        throw Error(
+          `Unexpected response, found more than one ${rel} Link headers.`,
+        );
+      }
+      if (link.length === 0) {
+        return;
+      }
+      result[rel] = toCredentialFilter(new URL(link[0].uri));
+    });
+  }
+  const body = await response.json();
+  if (!hasItems(body) || !body.items.every(isObjectWithId)) {
+    throw Error(
+      `Unexpected response, no items found in ${JSON.stringify(body)}`,
+    );
+  }
+  result.items = await Promise.all(
+    body.items.map((vc) => {
+      return verifiableCredentialToDataset(vc);
+    }),
+  );
+  return result;
+}
+
+function toQueryUrl(endpoint: URL, filter: CredentialFilter): URL {
+  const result = new URL(endpoint);
+  Object.entries(filter).forEach(([key, value]) => {
+    if (isSupportedFilterElement(key)) {
+      result.searchParams.append(key, value.toString());
+    }
+  });
+  return result;
+}
+
+/**
+ * Query for Access Requests or Access Grants based on a given filter.
+ *
+ * @param filter the query filter
+ * @param options query options
+ * @returns a paginated set of Access Credential matching the given filter
+ * @since unreleased
+ *
+ * @example
+ * ```
+ *  // Get the first results page.
+ *  const activeGrantsToday = await query(
+ *       {
+ *         type: "SolidAccessGrant",
+ *         status: "Active",
+ *         issuedWithin: DURATION.ONE_DAY,
+ *       },
+ *       {
+ *         fetch: session.fetch,
+ *         queryEndpoint: config.queryEndpoint,
+ *       },
+ *     );
+ * // Get the next results page.
+ * const activeGrantsToday2 = await query(
+ *       activeGrantsToday.next,
+ *       {
+ *         fetch: session.fetch,
+ *         queryEndpoint: config.queryEndpoint,
+ *       },
+ *     );
+ * ```
+ */
+export async function query(
```

Comment:
This is the main addition to the public API. The rest of this file is not exported.